src/soc/intel/baytrail/txei.c

```diff
@@ -0,0 +1,1196 @@
+
+		if (check_heci_reset())
+			txe_txei_init();
+
+		if (!wait_heci_ipc_ready()) {
+			printk(BIOS_ERR, "HECI: not ready\n");
+			break;
+		}
+
+		/*
+		 * Receive multiple packets until we meet one marked
+		 * complete or we run out of space in caller-provided buffer.
+		 */
+		do {
+			ret = recv_one_message(&hdr, p, left, &received);
+			if (ret) {
+				printk(BIOS_ERR, "HECI: Failed to receive!\n");
+				goto TXE_RX_ERR_HANDLE;
+			}
+			left -= received;
+			p += received;
+			/* If we read out everything ping to send more */
+			if (!(hdr & TXEI_HDR_IS_COMPLETE) && !txe_filled_slots())
+				host_gen_interrupt();
+		} while (received && !(hdr & TXEI_HDR_IS_COMPLETE) && left > 0);
+
+		if ((hdr & TXEI_HDR_IS_COMPLETE) && received) {
+			*maxlen = p - (uint8_t *)buff;
+			return TXE_TX_RX_SUCCESS;
+		}
+	}
+
+TXE_RX_ERR_HANDLE:
+	printk(BIOS_DEBUG, "HECI: Trigger HECI Reset\n");
+	heci_reset();
+	return TXE_RX_ERR_TXE_NOT_READY;
+}
+
+static enum txe_tx_rx_status heci_send_receive(const void *snd_msg, size_t snd_sz, void *rcv_msg,
+					size_t *rcv_sz, uint8_t txe_addr)
+{
+	enum txe_tx_rx_status ret;
+
+	ret = heci_send(snd_msg, snd_sz, BIOS_HOST_ADDR, txe_addr);
+	if (ret) {
+		printk(BIOS_ERR, "HECI: send Failed\n");
+		return ret;
+	}
+
+	if (rcv_msg != NULL) {
+		ret = heci_receive(rcv_msg, rcv_sz);
+		if (ret) {
+			printk(BIOS_ERR, "HECI: receive Failed\n");
+			return ret;
+		}
+	}
+	return ret;
+}
+
+static bool txe_cws_normal(void)
+{
+	if (!is_txe_devfn_visible())
+		return false;
+
+	uint32_t reg = pci_read_config32(PCI_DEV(0, TXE_DEV, TXE_FUNC), TXE_FWSTS0);
+
+	return ((reg & TXE_CWS_MASK) == TXE_FWSTS0_CWS_NORMAL);
+}
+
+static bool txe_com_normal(void)
+{
+	if (!is_txe_devfn_visible())
+		return false;
+
+	uint32_t reg = pci_read_config32(PCI_DEV(0, TXE_DEV, TXE_FUNC), TXE_FWSTS0);
+	reg &= TXE_CURR_OPMODE_MASK;
+	reg >>= TXE_CURR_OPMODE_SHIFT;
+
+	return (reg == TXE_FWSTS0_COM_NORMAL);
+}
+
+static bool txe_did_done(void)
+{
+	if (!is_txe_devfn_visible())
+		return false;
+
+	uint32_t reg = pci_read_config32(PCI_DEV(0, TXE_DEV, TXE_FUNC), TXE_DID_MSG);
+
+	return ((reg & DID_MSG_DID_DONE) == DID_MSG_DID_DONE);
+}
+
+int txe_global_reset(void)
+{
+	int status;
+	struct mkhi_hdr reply;
+	struct reset_message {
+		struct mkhi_hdr hdr;
+		uint8_t req_origin;
+		uint8_t reset_type;
+	} __packed;
+	struct reset_message msg = {
+		.hdr = {
+			.group_id = MKHI_CBM_GROUP_ID,
+			.command = MKHI_CBM_GLOBAL_RESET,
+		},
+		.req_origin = GR_ORIGIN_BIOS_POST,
+		.reset_type = 1
+	};
+	size_t reply_size;
+
+	printk(BIOS_DEBUG, "HECI: Global Reset Command\n");
+
+	if (!txe_cws_normal()) {
+		printk(BIOS_ERR, "HECI: TXE does not meet required prerequisites\n");
+		return 0;
+	}
+
+	heci_reset();
+
+	reply_size = sizeof(reply);
+	memset(&reply, 0, reply_size);
+
+	status = heci_send_receive(&msg, sizeof(msg), &reply, &reply_size, HECI_MKHI_ADDR);
+
+	printk(BIOS_DEBUG, "HECI: Global Reset %s!\n", !status ? "success" : "failure");
```

Comment:
In case of success HECI will reset the platform, right? In that case it doesn't make sense to print anything, it would be best to loop quietly until the reset happens.

Reply:
Common ME code doesn't do it: https://github.com/coreboot/coreboot/blob/master/src/soc/intel/common/block/cse/cse.c#L727
and I won't either